### PR TITLE
HPCC-15525 Ensure index parts are not reloaded in child queries

### PR DESCRIPTION
**Pull request by jakesmith**
This commit also tidies up some of the common handling of index
parts.

Signed-off-by: Jake Smith jake.smith@lexisnexis.com

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: thorlcr/activities/indexread/thindexreadslave.cpp [hunk 1]
```diff
@@ -152,17 +127,72 @@ class CIndexReadSlaveBase : public CSlaveActivity
...
-
+    IKeyIndex *openKeyPart(CActivityBase *activity, IPartDescriptor &partDesc)
+    {
+        RemoteFilename rfn;
+        partDesc.getFilename(0, rfn);
+        StringBuffer filePath;
+        rfn.getPath(filePath);
+        unsigned crc=0;
+        partDesc.getCrc(crc);
+        Owned<IDelayedFile> lfile = queryThor().queryFileCache().lookup(*activity, partDesc);
```
Comment (ghalliday) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: minor:  Now this is a member function activity doesn't need to be passed in.

File: thorlcr/activities/indexread/thindexreadslave.cpp [hunk 2]
```diff
@@ -630,27 +503,30 @@ class CIndexReadSlaveActivity : public CIndexReadSlaveBase
...
+            }
         }
         else if ((keyedLimit != RCMAX && (keyedLimitSkips || (helper->getFlags() & TIRcountkeyedlimit) != 0)))
-            keyedLimitCount = 0;            
+        {
+            eoi = false;
+            keyedLimitCount = 0;
+        }
         else
             eoi = true; // otherwise delayed until calc. in nextRow()
```
Comment (ghalliday) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: trivial: I think this comment i probably out of date.  I'm not sure what it means.

**Comment by ghalliday** [comment]
unrelated.  assert on line 832 (new code) is a bit strange.

**Comment by ghalliday** [comment]
@jakesmith some very minor comments.  Looks much simpler and cleaner.  I would merge as it stands.

**Comment by jakesmith** [comment]
>unrelated. assert on line 832 (new code) is a bit strange.

Very strange!
I presume that code have never been never hit, i.e. queryRowMetaData()->getMinRecordSize() is never = 1
If it were it would surely hit this assert (unless bizarrely chooseNLimit was = 1.
It's probably supposed to say:
assertex(choosenLimit < 256) or similar.

File: thorlcr/activities/indexread/thindexreadslave.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: confusing at least - I will remove

File: thorlcr/activities/indexread/thindexreadslave.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: Agree, will change.

**Comment by jakesmith** [comment]
@ghalliday - please see minor changes (please check the change to strange assert)

**Comment by ghalliday** [comment]
The strange assert would have been ok (just logically wrong).  I think getMinSize() ==1 is only true for exists(), but that would also happen to set the limit to 1...

**Pull request merged**